sycl/doc/EnvironmentVariables.md

```diff
@@ -11,6 +11,7 @@ subject to change. Do not rely on these variables in production code.
 
 | Environment variable | Values | Description |
 | -------------------- | ------ | ----------- |
+| SYCL_DEVICE_TRIPLE | device_type[:backend:device_num] | Force SYCL RT to use the specified device with optional backend and device_num information. It can list any number of triples separated by commas. This env var affects all different device selectors and device discovery. When no device that can satisfy the triple, a heuristic will choose the device that has the closest match. Possible values of device_type are *,host,cpu,gpu,acc,*. Possible values of backend are opencl, level0, cuda. Device_num is an unsigned integer that indexes the enumeration of devices from get_device() call. Among triple entries, only the device_type is required. The other two entries, backend and device_num, are optional. For example, to use cpu and level0 gpu, one can set SYCL_DEVICE_TYPE=cpu,gpu:level0 |
```

Comment:
It's not clear to me from this description whether `SYCL_DEVICE_TRIPLE` affects only the standard device selectors or whether it affects all the devices that are made available to the application.  For example, what does this mean:

```
SYCL_DEVICE_TRIPLE="gpu:opencl"
```

Does it mean:

1) Identify all of the devices on the system regardless of their backend and make them all available to the application via APIs like `device::get_devices()`.  The standard selectors `gpu_selector` and `default_selector` will prefer to return the OpenCL GPU device over GPU devices from different backends, however they may return a different GPU device if the OpenCL one is not available.  Or,

2) The only device (other than the host device) made available to the application is the OpenCL GPU device.  No CPU or accelerator devices are made available to the application, and GPU devices from other backends are also not available.  Unavailable devices are **not** exposed to the application via APIs like `device::get_devices()` or via device selectors.

The current implementation does not seem to implement either of these options, so I'm not sure which is intended.

Reply:
Two other comments:

* Why list "host" as an option for `device_type`?  You say elsewhere that the host device is always available (and the spec requires this).  If the envvar is unable to disable the host device, why let the user specify it?

* The purpose of `device_num` is very unclear to me, and the implementation in this PR doesn't use it, so I can't tell from the implementation.  Can you clarify what it means?

Reply:
> It's not clear to me from this description whether `SYCL_DEVICE_TRIPLE` affects only the standard device selectors or whether it affects all the devices that are made available to the application. For example, what does this mean:
> 
> ```
> SYCL_DEVICE_TRIPLE="gpu:opencl"
> ```
> 
> Does it mean:
> 
> 1. Identify all of the devices on the system regardless of their backend and make them all available to the application via APIs like `device::get_devices()`.  The standard selectors `gpu_selector` and `default_selector` will prefer to return the OpenCL GPU device over GPU devices from different backends, however they may return a different GPU device if the OpenCL one is not available.  Or,
> 2. The only device (other than the host device) made available to the application is the OpenCL GPU device.  No CPU or accelerator devices are made available to the application, and GPU devices from other backends are also not available.  Unavailable devices are **not** exposed to the application via APIs like `device::get_devices()` or via device selectors.
> 
> The current implementation does not seem to implement either of these options, so I'm not sure which is intended.

I agreed to hold this WIP PR until I discuss further details with other people.
In the meantime, I will try to answer as far as I know.
Please note that my answer to your question may change in the end.

Currently, we agree that this new env var will be used to limit only relevant plugins will be loaded into SYCL RT.
sycl::platform and sycl::device are made available from loading plugins that support various those kinds.
So, if the user specified this env var "gpu:opencl", SYCL RT will only load opencl plugin, but not level_zero or cuda plugins.
Note that "gpu:opencl" is different from "*:opencl".
The former means the user wants only OpenCL GPU devices, and nothing else.
The latter means the user wants all different types of devices supported from OpenCL.
CPU or ACC devices won't be discovered due to this filter "gpu:opencl" by all possible device selectors.
Even the user-defined device_selector is limited to only OpenCL GPU.

Note that there were some user complaints about this filtering by env vars. (e.g., SYCL_BE).
I think this issue need to be clarified much more to avoid this issue in the future.

Again, this is still work in progress, and it could change as discuss further details.

Reply:
> Two other comments:
> 
> * Why list "host" as an option for `device_type`?  You say elsewhere that the host device is always available (and the spec requires this).  If the envvar is unable to disable the host device, why let the user specify it?
> * The purpose of `device_num` is very unclear to me, and the implementation in this PR doesn't use it, so I can't tell from the implementation.  Can you clarify what it means?

The reason why we allow "host" to be included in the device_type choices is that we may give some impression that it is not possible to select a host device if we do not include "host" as an option.

Device_num is used only when there are multiple devices that can satisfy "device_type:backend" and the user wants to point out which one to be selected by the device_selector. For example, consider that a system has two discrete GPU devices plugged in the PCIe slots.
In the last discussion, we wanted to use the index of the return vector of devices from platform::get_devices() as the "device_num", and use the device_num to allocate different workloads for each device. 

Please note that this kind of details are still yet to be discussed and spec'd.